ipc_common.c

```diff
@@ -1,77 +1,170 @@
 #include "ipc_common.h"
 
-long int find_file_size(FILE *fd) {
-	long int size = 0;
-	// remember current position
-	long int offset = ftell(fd);
-	// set fd to the end of the file
-	if (fseek(fd, 0L, SEEK_END) != 0) {
-		perror("fseek");
-		exit(EXIT_FAILURE);
+// check if the string only contains number
+// convert it to an integer
+int get_int(char *str_ptr)
+{
+	char *ptr = str_ptr;
+	while (*ptr != '\0')
+	{
+		if (*ptr < 48 || *ptr > 57)
+		{
+			printf("Please give a valid size of shared memory. Integer only.\n");
+			usage(RECEIVE_USAGE, 0);
+		}
+		ptr++;
 	}
-	// calculate size
-	size = ftell(fd);
-	// set back to the old position
-	if (fseek(fd, offset, SEEK_SET) != 0) {
-		perror("fseek");
-		exit(EXIT_FAILURE);
+	int size = atoi(str_ptr);
+	if (size <= 0)
+	{
+		printf("Please give a valid size of shared memory (> 0).\n");
+		usage(RECEIVE_USAGE, 0);
 	}
 	return size;
 }
 
-void receive_msg(char *server_ptr, FILE *fd) {
+long int find_file_size(int fd)
+{
+	struct stat64 stat;
+	fstat64(fd, &stat);
+	return (long int) stat.st_size;
+}
+
+shm_data_t* get_shared_memory_pointer(char *shm_name, unsigned num_retries, int shm_size)
+{
+	unsigned tries;
+	shm_data_t *ptr;
+	int fd;
+
+	printf("Trying to open the shared memory named %s\n", shm_name);
+	for (tries = 1;;)
+	{
+		fd = shm_open(shm_name, O_RDWR, 0660);
+		if (fd != -1)
+			break;
+		printf("Tried %ld time(s)\n", tries);
+		++tries;
+		if (tries > num_retries)
+		{
+			perror("shmn_open");
+			return MAP_FAILED;
+		}
+		sleep(1);
+	}
+
+	printf("Trying to map the shared memory object at [/dev/shmem%s].\n", shm_name);
+	for (tries = 1;;)
+	{
+		ptr = (shm_data_t*)mmap64(NULL, shm_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
```

Comment:
You limited the size of the shm at 1024KB so unsing mmap64 is overhead.

Reply:
I use mmap64 just in case I want to increase the limit.
I tried ipc_receivefile with exist file on my machine and can't replicate the situation.

Reply:
> I use mmap64 just in case I want to increase the limit. 

I think you should already increase the limit as the user can choose the size.


> I tried ipc_receivefile with exist file on my machine and can't replicate the situation.

Did you try with a larger file than expected ?

Reply:
I add a line to remove old files if any before writing. Could you try again?